File: python/src/abslang/evaluators/adapters/google.py
"""
Google Vertex AI Evaluation adapter.

Routes ABS LLM-based evaluations through the Vertex AI Gen AI Evaluation service
using the classic ``vertexai.evaluation`` SDK (EvalTask + PointwiseMetric) in
bring-your-own-response (BYOR) mode: ABS already has the response, so we evaluate a
single row without model inference.

Scores are normalized to 0-1 before returning; the runner applies ``threshold``.
"""

import asyncio
import os
import re
from typing import Any

from .. import ObservedStep, EvalResult
from ..trace_utils import resolve_ref, trace_to_text


# ── Metric mapping ──

# ABS type -> MetricPromptTemplateExamples.Pointwise attribute name
_POINTWISE_ATTR = {
    "Groundedness": "GROUNDEDNESS",
    "Relevance": "QUESTION_ANSWERING_QUALITY",
    "Coherence": "COHERENCE",
    "Fluency": "FLUENCY",
    # safety types all map to the single SAFETY metric (0=unsafe, 1=safe)
    "HateUnfairness": "SAFETY",
    "Violence": "SAFETY",
    "Sexual": "SAFETY",
    "SelfHarm": "SAFETY",
}

SAFETY_TYPES = {"HateUnfairness", "Violence", "Sexual", "SelfHarm"}


# ── Config ──

def configure(
    project: str | None = None,
    location: str | None = None,
    credentials: str | None = None,
) -> None:
    """Prepare the Google adapter. All values default to environment variables."""
    if project:
        os.environ["GOOGLE_CLOUD_PROJECT"] = project
    if location:
        os.environ["GOOGLE_CLOUD_LOCATION"] = location
    if credentials:
        os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = credentials


def _get_project() -> str | None:
    return os.environ.get("GOOGLE_CLOUD_PROJECT")


def _get_location() -> str:
    return os.environ.get("GOOGLE_CLOUD_LOCATION", "us-central1")


def _sdk_available() -> bool:
    try:
        import vertexai  # noqa: F401

        return True
    except Exception:
        return False


# ── Helpers ──

def _last_assistant_content(trace: list[ObservedStep]) -> str:
    for s in reversed(trace):
        if s.actor == "assistant" and s.content:
            return str(s.content)
    return ""


def _scale_max(rating_scale: Any) -> int:
    s = str(rating_scale)
    m = re.search(r"(\d+)\s*$", s)
    if m:
        return int(m.group(1))
    return 5


def _normalize(value: float, scale_max: int) -> float:
    v = float(value)
    if scale_max > 1:
        v = v / scale_max
    return max(0.0, min(1.0, v))


def _not_configured(type_name: str) -> EvalResult:
    return EvalResult(
        type=type_name,
        passed=False,
        score=0.0,
        reason=(
            "Google Vertex AI Evaluation is not configured. Set it up:\n"
            "  pip install 'abslang[google]'\n"
            "  export GOOGLE_CLOUD_PROJECT=your-project\n"
            "  export GOOGLE_CLOUD_LOCATION=us-central1\n"
            "  gcloud auth application-default login   # or GOOGLE_APPLICATION_CREDENTIALS\n"
            f"Then: abslang run session.abs.yaml --agent $URL --adapter {type_name}=google"
        ),
    )


def _build_managed_metric(abs_type: str) -> tuple[Any, str]:
    """Return (metric, metric_name) for a managed pointwise metric."""
    from vertexai.evaluation import MetricPromptTemplateExamples

    attr = _POINTWISE_ATTR[abs_type]
    metric = getattr(MetricPromptTemplateExamples.Pointwise, attr)
    if isinstance(metric, str):
        return metric, metric
    name = getattr(metric, "metric", None) or getattr(metric, "metric_name", None) or attr.lower()
    return metric, name


def _build_custom_metric(name: str, criteria: str, scale_max: int) -> tuple[Any, str]:
    from vertexai.evaluation import PointwiseMetric, PointwiseMetricPromptTemplate

    rubric = {str(scale_max): "Excellent — fully meets the criteria."}
    if scale_max > 1:
        rubric[str(scale_max - 1)] = "Good — mostly meets the criteria."
        rubric[str(max(1, scale_max // 2))] = "Partial — meets some criteria."
    rubric["1"] = "Poor — does not meet the criteria."

    template = PointwiseMetricPromptTemplate(
        criteria={"criteria": criteria},
        rating_rubric=rubric,
        input_variables=["prompt", "response"],
    )
    metric = PointwiseMetric(metric=name, metric_prompt_template=template)
    return metric, name


def _extract(result: Any, metric_name: str) -> tuple[float, str]:
    table = getattr(result, "metrics_table", None)
    score: float | None = None
    reason = ""

    if table is not None:
        score_col = f"{metric_name}/score"
        expl_col = f"{metric_name}/explanation"
        try:
            if score_col in table.columns:
                score = float(table.iloc[0][score_col])
            if expl_col in table.columns:
                reason = str(table.iloc[0][expl_col])
        except Exception:
            pass
        if score is None:
            for col in table.columns:
                if col.endswith("/score"):
                    try:
                        score = float(table.iloc[0][col])
                        break
                    except Exception:
                        pass

    if score is None:
        summary = getattr(result, "summary_metrics", {})
        mean_key = f"{metric_name}/mean"
        if summary and mean_key in summary:
            score = float(summary[mean_key])

    if score is None:
        score = 0.0
    return score, reason


def _evaluate_single(metric: Any, variables: dict[str, str]) -> tuple[float, str]:
    """Run a single-row BYOR evaluation. Executed in a thread by the caller."""
    import pandas as pd
    from vertexai.evaluation import EvalTask

    df = pd.DataFrame([variables])
    result = EvalTask(dataset=df, metrics=[metric]).evaluate()
    metric_name = _metric_name_of(metric)
    return _extract(result, metric_name)


def _metric_name_of(metric: Any) -> str:
    if isinstance(metric, str):
        return metric
    return getattr(metric, "metric", None) or getattr(metric, "metric_name", None) or "metric"


# ── Main adapter ──

async def google_adapter(
    trace: list[ObservedStep],
    evaluation: dict[str, Any],
) -> EvalResult:
    eval_type = evaluation["type"]

    if eval_type == "llm_judge":
        return await _llm_judge(trace, evaluation)
    if eval_type == "custom":
        return await _custom(trace, evaluation)
    if eval_type in _POINTWISE_ATTR:
        return await _managed(trace, evaluation, eval_type)

    return EvalResult(
        type=eval_type,
        passed=False,
        score=0.0,
        reason=f"Google adapter does not support evaluator type: {eval_type}",
    )


def _resolve_inputs(trace: list[ObservedStep], evaluation: dict[str, Any]) -> dict[str, str]:
    self_content = _last_assistant_content(trace)
    response = resolve_ref(trace, evaluation.get("response"), self_content) or self_content
    query = resolve_ref(trace, evaluation.get("query"), self_content)
    context = resolve_ref(trace, evaluation.get("context"), self_content)
    if not query:
        query = trace_to_text(trace)
    return {"prompt": query, "context": context, "response": response}


async def _managed(trace: list[ObservedStep], evaluation: dict[str, Any], eval_type: str) -> EvalResult:
    if not _sdk_available():
        return EvalResult(type=eval_type, passed=False, score=0.0,
                          reason="google-cloud-aiplatform is not installed. Run: pip install 'abslang[google]'")
    if not _get_project():
        return _not_configured(eval_type)

    try:
        metric, metric_name = _build_managed_metric(eval_type)
        variables = _resolve_inputs(trace, evaluation)
        score, reason = await asyncio.to_thread(_evaluate_single, metric, variables)
        # SAFETY returns 0=unsafe, 1=safe — already 0-1 and "higher = safer".
        # Other managed metrics are 0-1 (passing rate) — clamp defensively.
        score = max(0.0, min(1.0, score))
        threshold = evaluation.get("threshold", 0.5)
        return EvalResult(
            type=eval_type,
            passed=score >= threshold,
            score=score,
            reason=f"[{metric_name}] {reason}" if reason else f"[{metric_name}] score {score:.2f}",
        )
    except Exception as e:
        return EvalResult(type=eval_type, passed=False, score=0.0, reason=f"Google {eval_type} error: {e}")


async def _llm_judge(trace: list[ObservedStep], evaluation: dict[str, Any]) -> EvalResult:
    if not _sdk_available():
        return EvalResult(type="llm_judge", passed=False, score=0.0,
                          reason="google-cloud-aiplatform is not installed. Run: pip install 'abslang[google]'")
    if not _get_project():
        return _not_configured("llm_judge")

    criteria = evaluation.get("criteria") or evaluation.get("prompt") or "Is the response helpful and accurate?"
    scale_max = _scale_max(evaluation.get("rating_scale", "1-5"))

    try:
        metric, metric_name = _build_custom_metric("llm_judge", criteria, scale_max)
        variables = _resolve_inputs(trace, evaluation)
        raw_score, reason = await asyncio.to_thread(_evaluate_single, metric, variables)
        score = _normalize(raw_score, scale_max)
        threshold = evaluation.get("threshold", 0.5)
        return EvalResult(
            type="llm_judge",
            passed=score >= threshold,
            score=score,
            reason=f"[{metric_name}] {reason}" if reason else f"[{metric_name}] score {score:.2f}",
        )
    except Exception as e:
        return EvalResult(type="llm_judge", passed=False, score=0.0, reason=f"Google judge error: {e}")


async def _custom(trace: list[ObservedStep], evaluation: dict[str, Any]) -> EvalResult:
    if not _sdk_available():
        return EvalResult(type="custom", passed=False, score=0.0,
                          reason="google-cloud-aiplatform is not installed. Run: pip install 'abslang[google]'")
    if not _get_project():
        return _not_configured("custom")

    eid = evaluation.get("id", "custom")
    criteria = evaluation.get("criteria") or evaluation.get("prompt")
    if not criteria:
        return EvalResult(type="custom", passed=False, score=0.0,
                          reason=f"Google custom evaluator '{eid}' requires a 'criteria' or 'prompt' field.")
    scale_max = _scale_max(evaluation.get("rating_scale", "1-5"))

    try:
        metric, metric_name = _build_custom_metric(eid, criteria, scale_max)
        variables = _resolve_inputs(trace, evaluation)
        raw_score, reason = await asyncio.to_thread(_evaluate_single, metric, variables)
        score = _normalize(raw_score, scale_max)
        threshold = evaluation.get("threshold", 0.5)
        return EvalResult(
            type="custom",
            passed=score >= threshold,
            score=score,
            reason=f"[{eid}] {reason}" if reason else f"[{eid}] score {score:.2f}",
        )
    except Exception as e:
        return EvalResult(type="custom", passed=False, score=0.0, reason=f"Google {eid} error: {e}")

File: typescript/src/evaluators/adapters/google.ts
/**
 * Google Vertex AI adapter — local, npm-native (bring-your-own-response).
 *
 * There is no JavaScript port of the `vertexai.evaluation` Python SDK, so this
 * adapter renders the official pointwise metric prompt templates (Apache-2.0,
 * from `google-cloud-aiplatform` 2.1.3 — see google_prompts.ts) and calls a
 * Gemini model through the official `@google-cloud/vertexai` Node SDK. The
 * adapter never re-runs the agent: it evaluates the ABS trace it receives.
 *
 * Requires the optional `@google-cloud/vertexai` package:
 *
 *   npm install @google-cloud/vertexai
 *
 * Environment:
 *   GOOGLE_CLOUD_PROJECT         required
 *   GOOGLE_CLOUD_LOCATION        default us-central1
 *   GOOGLE_APPLICATION_CREDENTIALS  optional service-account JSON
 *   GOOGLE_EVALUATOR_MODEL       optional judge model (default gemini-2.0-flash-001)
 *
 * Scores are normalized to 0–1 before returning; the runner applies `threshold`.
 * Mirrors python/src/abslang/evaluators/adapters/google.py.
 */

import { EvalResult, ObservedStep } from "../builtin";
import { resolveRef, toText, traceToText } from "../trace_utils";
import { GOOGLE_PROMPTS } from "./google_prompts";

export const DEFAULT_MODEL_ID = "gemini-2.0-flash-001";

const SAFETY_TYPES = new Set(["HateUnfairness", "Violence", "Sexual", "SelfHarm"]);

// ABS type -> official pointwise template
const POINTWISE_TEMPLATES: Record<string, string> = {
  Groundedness: "Groundedness",
  Relevance: "QuestionAnsweringQuality",
  Coherence: "Coherence",
  Fluency: "Fluency",
  HateUnfairness: "Safety",
  Violence: "Safety",
  Sexual: "Safety",
  SelfHarm: "Safety",
};

// ── Config ──

export interface GoogleAdapterConfig {
  project?: string;
  location?: string;
  credentials?: string;
  evaluatorModel?: string;
}

let vertexClient: any = null;

/** Prepare the Google adapter. All values default to environment variables. */
export function configureGoogle(config: GoogleAdapterConfig = {}): void {
  if (config.project) process.env.GOOGLE_CLOUD_PROJECT = config.project;
  if (config.location) process.env.GOOGLE_CLOUD_LOCATION = config.location;
  if (config.credentials) process.env.GOOGLE_APPLICATION_CREDENTIALS = config.credentials;
  if (config.evaluatorModel) process.env.GOOGLE_EVALUATOR_MODEL = config.evaluatorModel;
  vertexClient = null;
}

function getProject(): string | undefined {
  return process.env.GOOGLE_CLOUD_PROJECT;
}

function getLocation(): string {
  return process.env.GOOGLE_CLOUD_LOCATION || "us-central1";
}

function loadSdk(): any | null {
  try {
    return require("@google-cloud/vertexai");
  } catch {
    return null;
  }
}

function getClient(): any | null {
  if (vertexClient) return vertexClient;
  const sdk = loadSdk();
  if (!sdk?.VertexAI) return null;
  vertexClient = new sdk.VertexAI({ project: getProject(), location: getLocation() });
  return vertexClient;
}

function notConfigured(type: string): EvalResult {
  return {
    type,
    passed: false,
    score: 0,
    code: "adapter.not_configured",
    reason:
      "Google Vertex AI Evaluation is not configured. Set it up:\n" +
      "  npm install @google-cloud/vertexai\n" +
      "  export GOOGLE_CLOUD_PROJECT=your-project\n" +
      "  export GOOGLE_CLOUD_LOCATION=us-central1\n" +
      "  gcloud auth application-default login   # or GOOGLE_APPLICATION_CREDENTIALS\n" +
      `Then: abslang run session.abs.yaml --agent $URL --adapter ${type}=google`,
  };
}

function isNotConfiguredError(error: unknown): boolean {
  const message = String((error as any)?.message ?? error).toLowerCase();
  return (
    message.includes("not installed") ||
    message.includes("not configured") ||
    message.includes("could not load the default credentials")
  );
}

// ── Model call ──

async function generate(prompt: string): Promise<string> {
  if (!getProject()) throw new Error("not configured");
  const client = getClient();
  if (!client) throw new Error("Vertex AI SDK is not installed");
  const model = client.getGenerativeModel({
    model: process.env.GOOGLE_EVALUATOR_MODEL || DEFAULT_MODEL_ID,
  });
  const result = await model.generateContent(prompt);
  const response = result?.response;
  if (response && typeof response.text === "function") return response.text();
  const parts = response?.candidates?.[0]?.content?.parts ?? [];
  return parts.map((part: any) => (typeof part?.text === "string" ? part.text : "")).join("");
}

// ── Prompt building ──

function render(template: string, values: Record<string, string>): string {
  return template.replace(/\{\{(\w+)\}\}/g, (_: string, key: string) => values[key] ?? "");
}

function outputFormat(min: number, max: number): string {
  return (
    "\n# Output Format\n" +
    "Output a single JSON object with exactly these keys:\n" +
    '- "explanation": a string explaining your rating.\n' +
    `- "score": an integer between ${min} and ${max} according to the Rating Rubric.`
  );
}

function serializeDictInOrder(elements: Record<string, string>): string {
  return Object.keys(elements)
    .sort()
    .map((key) => `${key}: ${elements[key]}`)
    .join("\n");
}

const DEFAULT_INSTRUCTION =
  "You are an expert evaluator. Your task is to evaluate the quality of the responses generated by AI models. " +
  "We will provide you with the user prompt and an AI-generated responses.\n" +
  "You should first read the user input carefully for analyzing the task, and then evaluate the quality of the " +
  "responses based on the Criteria provided in the Evaluation section below.\n" +
  "You will assign the response a rating following the Rating Rubric and Evaluation Steps. Give step by step " +
  "explanations for your rating, and only choose ratings from the Rating Rubric.";

/** Build a pointwise metric prompt mirroring the SDK's PointwiseMetricPromptTemplate. */
export function buildCustomPrompt(criteria: string, prompt: string, response: string, scaleMax: number): string {
  const rubric: Record<string, string> = { [String(scaleMax)]: "Excellent — fully meets the criteria." };
  if (scaleMax > 1) {
    rubric[String(scaleMax - 1)] = "Good — mostly meets the criteria.";
    rubric[String(Math.max(1, Math.floor(scaleMax / 2)))] = "Partial — meets some criteria.";
  }
  rubric["1"] = "Poor — does not meet the criteria.";

  return [
    "# Instruction",
    DEFAULT_INSTRUCTION,
    "",
    "# Evaluation",
    "## Criteria",
    `criteria: ${criteria}`,
    "",
    "## Rating Rubric",
    serializeDictInOrder(rubric),
    "",
    "# User Inputs and AI-generated Response",
    "## User Inputs",
    "### Prompt",
    prompt,
    "",
    "## AI-generated Response",
    response,
    outputFormat(1, scaleMax),
  ].join("\n");
}

// ── Response parsing ──

export function parsePointwiseResponse(content: string): { score: number | null; reason: string } {
  const text = String(content ?? "");
  let data: any = null;
  try {
    data = JSON.parse(text);
  } catch {
    const match = text.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        data = JSON.parse(match[0]);
      } catch {
        /* fall through */
      }
    }
  }
  if (data && typeof data === "object") {
    const raw = data.score ?? data.rating ?? null;
    const score =
      typeof raw === "number" ? raw : raw !== null && !Number.isNaN(Number(raw)) ? Number(raw) : null;
    return {
      score,
      reason: typeof data.explanation === "string" ? data.explanation : typeof data.reason === "string" ? data.reason : "",
    };
  }
  const scoreMatch = text.match(/(?:Score|Rating):\s*(\d+(?:\.\d+)?)/i);
  const reasonMatch = text.match(/(?:Explanation|Reason):\s*(.+?)(?:\n|$)/i);
  return {
    score: scoreMatch ? Number(scoreMatch[1]) : null,
    reason: reasonMatch ? reasonMatch[1].trim() : text.slice(0, 200),
  };
}

// ── Inputs ──

function lastAssistantContent(trace: ObservedStep[]): string {
  for (let i = trace.length - 1; i >= 0; i--) {
    const step = trace[i];
    if (step.actor === "assistant" && step.content !== undefined && step.content !== null) {
      return toText(step.content);
    }
  }
  return "";
}

function resolveInputs(trace: ObservedStep[], evaluation: any): { prompt: string; response: string } {
  const selfContent = lastAssistantContent(trace);
  let prompt = resolveRef(trace, evaluation.query, selfContent);
  if (!prompt) prompt = traceToText(trace);
  const response = resolveRef(trace, evaluation.response, selfContent) || selfContent;
  return { prompt, response };
}

function scaleMax(ratingScale: any): number {
  const match = String(ratingScale).match(/(\d+)\s*$/);
  return match ? Number(match[1]) : 5;
}

// ── Evaluators ──

async function evaluateManaged(trace: ObservedStep[], evaluation: any, evalType: string): Promise<EvalResult> {
  if (!getProject()) return notConfigured(evalType);

  const templateName = POINTWISE_TEMPLATES[evalType];
  const { prompt, response } = resolveInputs(trace, evaluation);
  const isSafety = SAFETY_TYPES.has(evalType);
  const rendered =
    render(GOOGLE_PROMPTS[templateName], { prompt, response }) +
    outputFormat(isSafety ? 0 : 1, isSafety ? 1 : 5);
  const threshold = typeof evaluation.threshold === "number" ? evaluation.threshold : 0.5;

  try {
    const content = await generate(rendered);
    const parsed = parsePointwiseResponse(content);
    let score: number;
    if (parsed.score === null) {
      score = 0;
    } else if (isSafety) {
      // Official safety rubric is binary: 1 = safe, 0 = unsafe.
      score = Math.max(0, Math.min(1, parsed.score));
    } else {
      score = Math.max(0, Math.min(1, parsed.score / 5)); // 1-5 Likert
    }
    const reason = parsed.reason || `score ${score.toFixed(2)}`;
    return {
      type: evalType,
      passed: score >= threshold,
      score,
      reason: `[${templateName.toLowerCase()}] ${reason}`,
    };
  } catch (error: any) {
    if (isNotConfiguredError(error)) return notConfigured(evalType);
    return {
      type: evalType,
      passed: false,
      score: 0,
      code: "adapter.error",
      reason: `Google ${evalType} error: ${error.message}`,
    };
  }
}

async function evaluateLlmJudge(trace: ObservedStep[], evaluation: any): Promise<EvalResult> {
  if (!getProject()) return notConfigured("llm_judge");

  const criteria = evaluation.criteria || evaluation.prompt || "Is the response helpful and accurate?";
  const scale = scaleMax(evaluation.rating_scale ?? "1-5");
  const { prompt, response } = resolveInputs(trace, evaluation);
  const threshold = typeof evaluation.threshold === "number" ? evaluation.threshold : 0.5;

  try {
    const content = await generate(buildCustomPrompt(criteria, prompt, response, scale));
    const parsed = parsePointwiseResponse(content);
    const score = parsed.score === null ? 0 : Math.max(0, Math.min(1, parsed.score / scale));
    const reason = parsed.reason || `score ${score.toFixed(2)}`;
    return {
      type: "llm_judge",
      passed: score >= threshold,
      score,
      reason: `[llm_judge] ${reason}`,
    };
  } catch (error: any) {
    if (isNotConfiguredError(error)) return notConfigured("llm_judge");
    return {
      type: "llm_judge",
      passed: false,
      score: 0,
      code: "adapter.error",
      reason: `Google judge error: ${error.message}`,
    };
  }
}

async function evaluateCustom(trace: ObservedStep[], evaluation: any): Promise<EvalResult> {
  if (!getProject()) return notConfigured("custom");

  const eid = evaluation.id ?? "custom";
  const criteria = evaluation.criteria || evaluation.prompt;
  if (!criteria) {
    return {
      type: "custom",
      passed: false,
      score: 0,
      code: "evaluator.missing_input",
      reason: `Google custom evaluator '${eid}' requires a 'criteria' or 'prompt' field.`,
    };
  }
  const scale = scaleMax(evaluation.rating_scale ?? "1-5");
  const { prompt, response } = resolveInputs(trace, evaluation);
  const threshold = typeof evaluation.threshold === "number" ? evaluation.threshold : 0.5;

  try {
    const content = await generate(buildCustomPrompt(criteria, prompt, response, scale));
    const parsed = parsePointwiseResponse(content);
    const score = parsed.score === null ? 0 : Math.max(0, Math.min(1, parsed.score / scale));
    const reason = parsed.reason || `score ${score.toFixed(2)}`;
    return {
      type: "custom",
      passed: score >= threshold,
      score,
      reason: `[${eid}] ${reason}`,
    };
  } catch (error: any) {
    if (isNotConfiguredError(error)) return notConfigured("custom");
    return {
      type: "custom",
      passed: false,
      score: 0,
      code: "adapter.error",
      reason: `Google ${eid} error: ${error.message}`,
    };
  }
}

// ── Main adapter ──

async function routeGoogle(trace: ObservedStep[], evaluation: any): Promise<EvalResult> {
  const evalType = evaluation.type;
  if (evalType === "llm_judge") return evaluateLlmJudge(trace, evaluation);
  if (evalType === "custom") return evaluateCustom(trace, evaluation);
  if (POINTWISE_TEMPLATES[evalType]) return evaluateManaged(trace, evaluation, evalType);
  return {
    type: evalType,
    passed: false,
    score: 0,
    code: "adapter.unsupported_type",
    reason: `Google adapter does not support evaluator type: ${evalType}`,
  };
}

export async function googleAdapter(trace: ObservedStep[], evaluation: any): Promise<EvalResult> {
  const result = await routeGoogle(trace, evaluation);
  return { ...result, adapter: result.adapter ?? "google" };
}
